### ISSUE TITLE
TCP connect completion failure with AWS IoT

### ISSUE THREAD

**Issue by rosslannen** [eclipse/paho.mqtt.rust#119]
I am trying to connect to AWS IoT using the MQTT protocol, X508 client certificates, and on port 443. When I try to connect, I am getting `Error: [-1] TCP connect completion failure` printed. I have included the code, as well as the log output for it. Any help is greatly appreciated!

Code:
``` rust
use paho_mqtt as mqtt;

const SERVER_HOST: &str = "<STUFF>.iot.<LOCATION>.amazonaws.com";
const ROOT_CA_PATH: &str = "root-CA.crt";
const CERTIFICATE_PATH: &str = "thing1.cert.pem";
const PRIVATE_KEY_PATH: &str = "thing1.private.key";
const ALPN_PROTOCOLS: &[&str] = &["x-amzn-mqtt-ca"];

fn main() -> anyhow::Result<()> {
    env_logger::init();

    let uri = format!("ssl://{}:{}", SERVER_HOST, 443);

    let options = mqtt::CreateOptionsBuilder::new()
        .server_uri(&uri)
        .persistence(None)
        .finalize();

    let client = mqtt::Client::new(options)?;

    let ssl_options = mqtt::SslOptionsBuilder::new()
        .trust_store(ROOT_CA_PATH)?
        .key_store(CERTIFICATE_PATH)?
        .private_key(PRIVATE_KEY_PATH)?
        .alpn_protos(ALPN_PROTOCOLS)
        .enable_server_cert_auth(false)
        .verify(true)
        .finalize();

    let connection_options = mqtt::ConnectOptionsBuilder::new()
        .clean_session(true)
        .ssl_options(ssl_options)
        .finalize();

    client.connect(connection_options)?;

    Ok(())
}
```

Output:
```
[2021-04-20T19:57:43Z DEBUG paho_mqtt::async_client] Create options: CreateOptions { copts: MQTTAsync_createOptions { struct_id: [77, 81, 67, 79], struct_version: 2, sendWhileDisconnected: 0, maxBufferedMessages: 100, MQTTVersion: 0, allowDisconnectedSendAtAnyTime: 0, deleteOldestMessages: 0, restoreMessages: 1, persistQoS0: 1 }, server_uri: "ssl://a1yw9avly5y2bp-ats.iot.us-east-2.amazonaws.com:443", client_id: "", persistence: None, user_data: None }
[2021-04-20T19:57:43Z DEBUG paho_mqtt::async_client] Creating client with persistence: 1
=========================================================
                   Trace Output
Product name: Eclipse Paho Asynchronous MQTT C Client Library
Version: 1.3.8
Build level: 2021-04-20T16:19:06Z
OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
OpenSSL flags: compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -Wa,--noexecstack -g -O2 -fdebug-prefix-map=/build/openssl-Juj39H/openssl-1.1.1f=. -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_TLS_SECURITY_LEVEL=2 -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAESNI_ASM -DVPAES_ASM -DGHASH_ASM -
OpenSSL build timestamp: built on: Mon Mar 22 11:37:17 2021 UTC
OpenSSL platform: platform: debian-amd64
OpenSSL directory: OPENSSLDIR: "/usr/lib/ssl"
/proc/version: Linux version 5.6.0-1054-oem (buildd@lgw01-amd64-018) (gcc version 9.3.0 (Ubuntu 9.3.0-17ubuntu1~20.04)) #58-Ubuntu SMP Wed Apr 14 13:38:12 UTC 2021

=========================================================
19691231 170000.000 0 commands restored for client 
19691231 170000.000 0 queued messages restored for client 
[2021-04-20T19:57:43Z DEBUG paho_mqtt::async_client] AsyncClient handle: 0x5605fa07a440
[2021-04-20T19:57:43Z DEBUG paho_mqtt::async_client] Connecting handle: 0x5605fa07a440
[2021-04-20T19:57:43Z DEBUG paho_mqtt::async_client] Connect options: ConnectOptions { copts: MQTTAsync_connectOptions { struct_id: [77, 81, 84, 67], struct_version: 8, keepAliveInterval: 60, cleansession: 1, maxInflight: 65535, will: 0x0, username: 0x0, password: 0x0, connectTimeout: 30, retryInterval: 0, ssl: 0x5605fa07adf0, onSuccess: None, onFailure: None, context: 0x0, serverURIcount: 0, serverURIs: 0x0, MQTTVersion: 0, automaticReconnect: 0, minRetryInterval: 1, maxRetryInterval: 60, binarypwd: MQTTAsync_connectOptions__bindgen_ty_1 { len: 0, data: 0x0 }, cleanstart: 0, connectProperties: 0x0, willProperties: 0x0, onSuccess5: None, onFailure5: None, httpHeaders: 0x0, httpProxy: 0x0, httpsProxy: 0x0 }, data: ConnectOptionsData { will: None, ssl: Some(SslOptions { copts: MQTTAsync_SSLOptions { struct_id: [77, 81, 84, 83], struct_version: 5, trustStore: 0x5605fa07ab90, keyStore: 0x5605fa07ab70, privateKey: 0x5605fa07ab50, privateKeyPassword: 0x0, enabledCipherSuites: 0x0, enableServerCertAuth: 0, sslVersion: 0, verify: 1, CApath: 0x0, ssl_error_cb: None, ssl_error_context: 0x0, ssl_psk_cb: None, ssl_psk_context: 0x0, disableDefaultTrustStore: 0, protos: 0x5605fa07ace0, protos_len: 15 }, data: SslOptionsData { trust_store: "root-CA.crt", key_store: "thing1.cert.pem", private_key: "thing1.private.key", private_key_password: "", enabled_cipher_suites: "", ca_path: "", protos: [14, 120, 45, 97, 109, 122, 110, 45, 109, 113, 116, 116, 45, 99, 97] } }), user_name: None, password: None, server_uris: StringCollection { c_coll: [], c_mut_coll: [], data: StringCollectionData { coll: [] } }, props: None, will_props: None, http_headers: None, http_proxy: None, https_proxy: None } }
19691231 170000.000 Connecting to serverURI a1yw9avly5y2bp-ats.iot.us-east-2.amazonaws.com:443 with MQTT version 4
19691231 170000.000 New socket 3 for a1yw9avly5y2bp-ats.iot.us-east-2.amazonaws.com:443, port 443
19691231 170000.000 Connect pending
19691231 170000.000 m->c->connect_state = 1
19691231 170000.000 SSL cipher available: 0:TLS_AES_256_GCM_SHA384
19691231 170000.000 SSL cipher available: 1:TLS_CHACHA20_POLY1305_SHA256
19691231 170000.000 SSL cipher available: 2:TLS_AES_128_GCM_SHA256
19691231 170000.000 SSL cipher available: 3:ECDHE-ECDSA-AES256-GCM-SHA384
19691231 170000.000 SSL cipher available: 4:ECDHE-RSA-AES256-GCM-SHA384
19691231 170000.000 SSL cipher available: 5:DHE-RSA-AES256-GCM-SHA384
19691231 170000.000 SSL cipher available: 6:ECDHE-ECDSA-CHACHA20-POLY1305
19691231 170000.000 SSL cipher available: 7:ECDHE-RSA-CHACHA20-POLY1305
19691231 170000.000 SSL cipher available: 8:DHE-RSA-CHACHA20-POLY1305
19691231 170000.000 SSL cipher available: 9:ECDHE-ECDSA-AES128-GCM-SHA256
19691231 170000.000 SSL cipher available: 10:ECDHE-RSA-AES128-GCM-SHA256
19691231 170000.000 SSL cipher available: 11:DHE-RSA-AES128-GCM-SHA256
19691231 170000.000 SSL cipher available: 12:ECDHE-ECDSA-AES256-SHA384
20210420 135743.945 SSL cipher available: 13:ECDHE-RSA-AES256-SHA384
20210420 135743.945 SSL cipher available: 14:DHE-RSA-AES256-SHA256
20210420 135743.945 SSL cipher available: 15:ECDHE-ECDSA-AES128-SHA256
20210420 135743.945 SSL cipher available: 16:ECDHE-RSA-AES128-SHA256
20210420 135743.945 SSL cipher available: 17:DHE-RSA-AES128-SHA256
20210420 135743.945 SSL cipher available: 18:ECDHE-ECDSA-AES256-SHA
20210420 135743.945 SSL cipher available: 19:ECDHE-RSA-AES256-SHA
20210420 135743.945 SSL cipher available: 20:DHE-RSA-AES256-SHA
20210420 135743.945 SSL cipher available: 21:ECDHE-ECDSA-AES128-SHA
20210420 135743.945 SSL cipher available: 22:ECDHE-RSA-AES128-SHA
20210420 135743.945 SSL cipher available: 23:DHE-RSA-AES128-SHA
20210420 135743.945 SSL cipher available: 24:RSA-PSK-AES256-GCM-SHA384
20210420 135743.945 SSL cipher available: 25:DHE-PSK-AES256-GCM-SHA384
20210420 135743.945 SSL cipher available: 26:RSA-PSK-CHACHA20-POLY1305
20210420 135743.945 SSL cipher available: 27:DHE-PSK-CHACHA20-POLY1305
20210420 135743.945 SSL cipher available: 28:ECDHE-PSK-CHACHA20-POLY1305
20210420 135743.945 SSL cipher available: 29:AES256-GCM-SHA384
20210420 135743.945 SSL cipher available: 30:PSK-AES256-GCM-SHA384
20210420 135743.945 SSL cipher available: 31:PSK-CHACHA20-POLY1305
20210420 135743.945 SSL cipher available: 32:RSA-PSK-AES128-GCM-SHA256
20210420 135743.945 SSL cipher available: 33:DHE-PSK-AES128-GCM-SHA256
20210420 135743.945 SSL cipher available: 34:AES128-GCM-SHA256
20210420 135743.945 SSL cipher available: 35:PSK-AES128-GCM-SHA256
20210420 135743.945 SSL cipher available: 36:AES256-SHA256
20210420 135743.945 SSL cipher available: 37:AES128-SHA256
20210420 135743.945 SSL cipher available: 38:ECDHE-PSK-AES256-CBC-SHA384
20210420 135743.945 SSL cipher available: 39:ECDHE-PSK-AES256-CBC-SHA
20210420 135743.945 SSL cipher available: 40:SRP-RSA-AES-256-CBC-SHA
20210420 135743.945 SSL cipher available: 41:SRP-AES-256-CBC-SHA
20210420 135743.945 SSL cipher available: 42:RSA-PSK-AES256-CBC-SHA384
20210420 135743.945 SSL cipher available: 43:DHE-PSK-AES256-CBC-SHA384
20210420 135743.945 SSL cipher available: 44:RSA-PSK-AES256-CBC-SHA
20210420 135743.945 SSL cipher available: 45:DHE-PSK-AES256-CBC-SHA
20210420 135743.945 SSL cipher available: 46:AES256-SHA
20210420 135743.945 SSL cipher available: 47:PSK-AES256-CBC-SHA384
20210420 135743.945 SSL cipher available: 48:PSK-AES256-CBC-SHA
20210420 135743.945 SSL cipher available: 49:ECDHE-PSK-AES128-CBC-SHA256
20210420 135743.945 SSL cipher available: 50:ECDHE-PSK-AES128-CBC-SHA
20210420 135743.945 SSL cipher available: 51:SRP-RSA-AES-128-CBC-SHA
20210420 135743.945 SSL cipher available: 52:SRP-AES-128-CBC-SHA
20210420 135743.945 SSL cipher available: 53:RSA-PSK-AES128-CBC-SHA256
20210420 135743.945 SSL cipher available: 54:DHE-PSK-AES128-CBC-SHA256
20210420 135743.945 SSL cipher available: 55:RSA-PSK-AES128-CBC-SHA
20210420 135743.945 SSL cipher available: 56:DHE-PSK-AES128-CBC-SHA
20210420 135743.945 SSL cipher available: 57:AES128-SHA
20210420 135743.945 SSL cipher available: 58:PSK-AES128-CBC-SHA256
20210420 135743.945 SSL cipher available: 59:PSK-AES128-CBC-SHA
20210420 135743.945 the src string is truncated
20210420 135743.945 SSL handshake started write:unknown:unknown
20210420 135743.945 SSL state connect:before SSL initialization:(NONE)
20210420 135743.945 sent 0 256 buflen 5
20210420 135743.945 sent 772 22 buflen 354
20210420 135743.945 SSL state connect:SSLv3/TLS write client hello:(NONE)
20210420 135743.945 SSL connect:SSLv3/TLS write client hello
20210420 135743.945 SSLSocket error WANT_READ/WANT_WRITE
20210420 135743.945 m->c->connect_state = 2
20210420 135743.945 received 0 256 buflen 5
20210420 135743.945 SSL state connect:SSLv3/TLS write client hello:(NONE)
20210420 135743.945 received 772 22 buflen 112
20210420 135743.945 SSL connect:SSLv3/TLS read server hello
20210420 135744.021 SSLSocket error WANT_READ/WANT_WRITE
20210420 135744.021 m->c->connect_state = 2
20210420 135744.021 received 0 256 buflen 5
20210420 135744.021 SSL state connect:SSLv3/TLS read server hello:(NONE)
20210420 135744.021 received 771 22 buflen 4883
20210420 135744.021 received 0 256 buflen 5
20210420 135744.021 SSL state connect:SSLv3/TLS read server certificate:(NONE)
20210420 135744.021 received 771 22 buflen 333
20210420 135744.021 received 0 256 buflen 5
20210420 135744.021 SSL state connect:SSLv3/TLS read server key exchange:(NONE)
20210420 135744.021 received 771 22 buflen 32
20210420 135744.021 received 0 256 buflen 5
20210420 135744.021 SSL state connect:SSLv3/TLS read server certificate request:(NONE)
20210420 135744.021 received 771 22 buflen 4
20210420 135744.021 SSL state connect:SSLv3/TLS read server done:(NONE)
20210420 135744.021 sent 0 256 buflen 5
20210420 135744.021 sent 771 22 buflen 871
20210420 135744.021 SSL state connect:SSLv3/TLS write client certificate:(NONE)
20210420 135744.021 sent 0 256 buflen 5
20210420 135744.021 sent 771 22 buflen 70
20210420 135744.046 SSL state connect:SSLv3/TLS write client key exchange:(NONE)
20210420 135744.046 sent 0 256 buflen 5
20210420 135744.046 sent 771 22 buflen 264
20210420 135744.046 SSL state connect:SSLv3/TLS write certificate verify:(NONE)
20210420 135744.046 sent 0 256 buflen 5
20210420 135744.046 sent 771 20 buflen 1
20210420 135744.046 SSL state connect:SSLv3/TLS write change cipher spec:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.046 sent 0 256 buflen 5
20210420 135744.046 sent 771 22 buflen 16
20210420 135744.046 SSL state connect:SSLv3/TLS write finished:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.046 SSL connect:SSLv3/TLS write finished
20210420 135744.046 SSLSocket error WANT_READ/WANT_WRITE
20210420 135744.046 m->c->connect_state = 2
20210420 135744.046 received 0 256 buflen 5
20210420 135744.046 SSL state connect:SSLv3/TLS write finished:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.046 received 0 256 buflen 5
20210420 135744.046 SSL state connect:SSLv3/TLS read change cipher spec:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.046 received 771 22 buflen 16
20210420 135744.046 SSL state connect:SSLv3/TLS read finished:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.046 SSL handshake done write:unknown:unknown
20210420 135744.112 SSL certificate verification: X509_V_OK
20210420 135744.112 SSL connect:SSL negotiation finished successfully
20210420 135744.112 peername from X509_check_host is *.iot.us-east-2.amazonaws.com
20210420 135744.112 sent 0 256 buflen 5
20210420 135744.112 3  -> CONNECT version 4 clean: 1 (0)
20210420 135744.112 m->c->connect_state = 4
20210420 135744.112 received 0 256 buflen 5
20210420 135744.112 received 771 21 buflen 2
20210420 135744.112 SSL alert read:warning:close notify
20210420 135744.112 CONNECT sent but MQTTPacket_Factory has returned SOCKET_ERROR
20210420 135744.112 sent 0 256 buflen 5
20210420 135744.112 sent 771 21 buflen 2
20210420 135744.112 SSL alert write:warning:close notify
20210420 135744.112 Removed socket 3
20210420 135744.112 Connect failed, more to try
20210420 135744.112 Could not find client corresponding to socket 3
20210420 135744.112 Connecting to serverURI a1yw9avly5y2bp-ats.iot.us-east-2.amazonaws.com:443 with MQTT version 3
20210420 135744.112 New socket 3 for a1yw9avly5y2bp-ats.iot.us-east-2.amazonaws.com:443, port 443
20210420 135744.112 Connect pending
20210420 135744.112 m->c->connect_state = 1
20210420 135744.332 SSL cipher available: 0:TLS_AES_256_GCM_SHA384
20210420 135744.332 SSL cipher available: 1:TLS_CHACHA20_POLY1305_SHA256
20210420 135744.332 SSL cipher available: 2:TLS_AES_128_GCM_SHA256
20210420 135744.332 SSL cipher available: 3:ECDHE-ECDSA-AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 4:ECDHE-RSA-AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 5:DHE-RSA-AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 6:ECDHE-ECDSA-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 7:ECDHE-RSA-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 8:DHE-RSA-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 9:ECDHE-ECDSA-AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 10:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 11:DHE-RSA-AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 12:ECDHE-ECDSA-AES256-SHA384
20210420 135744.332 SSL cipher available: 13:ECDHE-RSA-AES256-SHA384
20210420 135744.332 SSL cipher available: 14:DHE-RSA-AES256-SHA256
20210420 135744.332 SSL cipher available: 15:ECDHE-ECDSA-AES128-SHA256
20210420 135744.332 SSL cipher available: 16:ECDHE-RSA-AES128-SHA256
20210420 135744.332 SSL cipher available: 17:DHE-RSA-AES128-SHA256
20210420 135744.332 SSL cipher available: 18:ECDHE-ECDSA-AES256-SHA
20210420 135744.332 SSL cipher available: 19:ECDHE-RSA-AES256-SHA
20210420 135744.332 SSL cipher available: 20:DHE-RSA-AES256-SHA
20210420 135744.332 SSL cipher available: 21:ECDHE-ECDSA-AES128-SHA
20210420 135744.332 SSL cipher available: 22:ECDHE-RSA-AES128-SHA
20210420 135744.332 SSL cipher available: 23:DHE-RSA-AES128-SHA
20210420 135744.332 SSL cipher available: 24:RSA-PSK-AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 25:DHE-PSK-AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 26:RSA-PSK-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 27:DHE-PSK-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 28:ECDHE-PSK-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 29:AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 30:PSK-AES256-GCM-SHA384
20210420 135744.332 SSL cipher available: 31:PSK-CHACHA20-POLY1305
20210420 135744.332 SSL cipher available: 32:RSA-PSK-AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 33:DHE-PSK-AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 34:AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 35:PSK-AES128-GCM-SHA256
20210420 135744.332 SSL cipher available: 36:AES256-SHA256
20210420 135744.332 SSL cipher available: 37:AES128-SHA256
20210420 135744.332 SSL cipher available: 38:ECDHE-PSK-AES256-CBC-SHA384
20210420 135744.332 SSL cipher available: 39:ECDHE-PSK-AES256-CBC-SHA
20210420 135744.332 SSL cipher available: 40:SRP-RSA-AES-256-CBC-SHA
20210420 135744.332 SSL cipher available: 41:SRP-AES-256-CBC-SHA
20210420 135744.332 SSL cipher available: 42:RSA-PSK-AES256-CBC-SHA384
20210420 135744.332 SSL cipher available: 43:DHE-PSK-AES256-CBC-SHA384
20210420 135744.332 SSL cipher available: 44:RSA-PSK-AES256-CBC-SHA
20210420 135744.332 SSL cipher available: 45:DHE-PSK-AES256-CBC-SHA
20210420 135744.332 SSL cipher available: 46:AES256-SHA
20210420 135744.332 SSL cipher available: 47:PSK-AES256-CBC-SHA384
20210420 135744.332 SSL cipher available: 48:PSK-AES256-CBC-SHA
20210420 135744.332 SSL cipher available: 49:ECDHE-PSK-AES128-CBC-SHA256
20210420 135744.332 SSL cipher available: 50:ECDHE-PSK-AES128-CBC-SHA
20210420 135744.332 SSL cipher available: 51:SRP-RSA-AES-128-CBC-SHA
20210420 135744.332 SSL cipher available: 52:SRP-AES-128-CBC-SHA
20210420 135744.332 SSL cipher available: 53:RSA-PSK-AES128-CBC-SHA256
20210420 135744.332 SSL cipher available: 54:DHE-PSK-AES128-CBC-SHA256
20210420 135744.332 SSL cipher available: 55:RSA-PSK-AES128-CBC-SHA
20210420 135744.332 SSL cipher available: 56:DHE-PSK-AES128-CBC-SHA
20210420 135744.332 SSL cipher available: 57:AES128-SHA
20210420 135744.332 SSL cipher available: 58:PSK-AES128-CBC-SHA256
20210420 135744.332 SSL cipher available: 59:PSK-AES128-CBC-SHA
20210420 135744.332 the src string is truncated
20210420 135744.332 SSL handshake started write:unknown:unknown
20210420 135744.332 SSL state connect:before SSL initialization:(NONE)
20210420 135744.332 sent 0 256 buflen 5
20210420 135744.332 sent 772 22 buflen 354
20210420 135744.332 SSL state connect:SSLv3/TLS write client hello:(NONE)
20210420 135744.332 SSL connect:SSLv3/TLS write client hello
20210420 135744.332 SSLSocket error WANT_READ/WANT_WRITE
20210420 135744.332 m->c->connect_state = 2
20210420 135744.332 received 0 256 buflen 5
20210420 135744.332 SSL state connect:SSLv3/TLS write client hello:(NONE)
20210420 135744.332 received 772 22 buflen 112
20210420 135744.332 received 0 256 buflen 5
20210420 135744.332 SSL state connect:SSLv3/TLS read server hello:(NONE)
20210420 135744.332 received 771 22 buflen 4883
20210420 135744.332 SSL connect:SSLv3/TLS read server certificate
20210420 135744.332 SSLSocket error WANT_READ/WANT_WRITE
20210420 135744.332 m->c->connect_state = 2
20210420 135744.332 received 0 256 buflen 5
20210420 135744.332 SSL state connect:SSLv3/TLS read server certificate:(NONE)
20210420 135744.410 received 771 22 buflen 333
20210420 135744.410 received 0 256 buflen 5
20210420 135744.410 SSL state connect:SSLv3/TLS read server key exchange:(NONE)
20210420 135744.410 received 771 22 buflen 32
20210420 135744.410 received 0 256 buflen 5
20210420 135744.410 SSL state connect:SSLv3/TLS read server certificate request:(NONE)
20210420 135744.410 received 771 22 buflen 4
20210420 135744.410 SSL state connect:SSLv3/TLS read server done:(NONE)
20210420 135744.410 sent 0 256 buflen 5
20210420 135744.410 sent 771 22 buflen 871
20210420 135744.410 SSL state connect:SSLv3/TLS write client certificate:(NONE)
20210420 135744.410 sent 0 256 buflen 5
20210420 135744.410 sent 771 22 buflen 70
20210420 135744.410 SSL state connect:SSLv3/TLS write client key exchange:(NONE)
20210420 135744.410 sent 0 256 buflen 5
20210420 135744.410 sent 771 22 buflen 264
20210420 135744.410 SSL state connect:SSLv3/TLS write certificate verify:(NONE)
20210420 135744.410 sent 0 256 buflen 5
20210420 135744.410 sent 771 20 buflen 1
20210420 135744.410 SSL state connect:SSLv3/TLS write change cipher spec:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.412 sent 0 256 buflen 5
20210420 135744.412 sent 771 22 buflen 16
20210420 135744.412 SSL state connect:SSLv3/TLS write finished:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.412 SSL connect:SSLv3/TLS write finished
20210420 135744.412 SSLSocket error WANT_READ/WANT_WRITE
20210420 135744.412 m->c->connect_state = 2
20210420 135744.412 received 0 256 buflen 5
20210420 135744.412 SSL state connect:SSLv3/TLS write finished:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.412 received 0 256 buflen 5
20210420 135744.412 SSL state connect:SSLv3/TLS read change cipher spec:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.412 received 771 22 buflen 16
20210420 135744.412 SSL state connect:SSLv3/TLS read finished:ECDHE-RSA-AES128-GCM-SHA256
20210420 135744.412 SSL handshake done write:unknown:unknown
20210420 135744.412 SSL certificate verification: X509_V_OK
20210420 135744.412 SSL connect:SSL negotiation finished successfully
20210420 135744.412 peername from X509_check_host is *.iot.us-east-2.amazonaws.com
20210420 135744.412 sent 0 256 buflen 5
20210420 135744.412 3  -> CONNECT version 3 clean: 1 (0)
20210420 135744.412 m->c->connect_state = 4
20210420 135744.412 received 0 256 buflen 5
20210420 135744.548 received 771 21 buflen 2
20210420 135744.548 SSL alert read:warning:close notify
20210420 135744.548 CONNECT sent but MQTTPacket_Factory has returned SOCKET_ERROR
20210420 135744.548 sent 0 256 buflen 5
20210420 135744.548 sent 771 21 buflen 2
20210420 135744.548 SSL alert write:warning:close notify
20210420 135744.548 Removed socket 3
20210420 135744.548 0 inflight messages deleted for client 
20210420 135744.548 0 responses removed for client 
20210420 135744.548 Calling connect failure for client 
[2021-04-20T19:57:44Z DEBUG paho_mqtt::token] Token failure! 0x5605fa07beb0, 0x7fbf66635ad0
[2021-04-20T19:57:44Z DEBUG paho_mqtt::token] Token failure message: "TCP connect completion failure"
[2021-04-20T19:57:44Z DEBUG paho_mqtt::token] Token completed with code: -1
[2021-04-20T19:57:44Z DEBUG paho_mqtt::token] Expecting server response for: Connect
[2021-04-20T19:57:44Z DEBUG paho_mqtt::token] Got response: ServerResponse { rsp: None, props: Properties { cprops: MQTTProperties { count: 0, max_count: 0, length: 0, array: 0x0 } }, reason_code: Success }
20210420 135744.548 Could not find client corresponding to socket 3
20210420 135744.548 0 inflight messages deleted for client 
20210420 135744.548 0 responses removed for client 
20210420 135744.548 0 responses removed for client 
20210420 135744.548 0 commands removed for client 
20210420 135744.548 Removed client  from bstate->clients, socket 0
20210420 135744.548 Conn_count is 0
20210420 135744.548 sleeping
Error: [-1] TCP connect completion failure
```

**Comment by rosslannen** [comment]
Update, this was an issue with the policy for the thing in IoT core, not with the client code running on the device. Link to the fix I used here: https://github.com/aws/aws-iot-device-sdk-python-v2/issues/148

**Comment by fpagliughi** [comment]
Did you get it to connect?  The ALPN stuff was reported to be working in the C client, but I haven't re-activated my AWS account to try it out myself in Rust.

BTW, I hope that eventually the C lib will bubble up better error messages, but to help analyze what you initially saw...  
The biggest problem with MQTT v3 is that on detecting an error, the server has no way to tell you what it doesn't like, and can only drop the connection with no error given. This is vastly improved in MQTT v5, but the big services like AWS aren't using it yet.

From the logs, the initial TLS connection gets established:
```
20210420 135744.412 SSL connect:SSL negotiation finished successfully
```
But then when the library send the MQTT  _CONNECT_ packet, the server just disconnects:
```
20210420 135744.412 3  -> CONNECT version 3 clean: 1 (0)
...
20210420 135744.548 SSL alert read:warning:close notify
20210420 135744.548 CONNECT sent but MQTTPacket_Factory has returned SOCKET_ERROR
```

So you're just left guessing as to what AWS didn't like!

**Comment by rosslannen** [comment]
I did get it to connect! It ended up being a permissions issue for the device in AWS IoT. The fix was to set the IoT policy to allow connecting/publishing/etc to all topics. The defaults for the tutorials apparently only allow very limited access to the specific things the tutorials try to accomplish. If you try to access something outside of those, AWS simply drops the connection. I ran into the same error while playing around with the official python v2 IoT SDK, at which point I figured it was not a problem with the client side of things and found the above linked issue.

Short answer, if you set the device policy to the below JSON, the above code works like a charm:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iot:*"
            ],
            "Resource": "*"
        }
    ]
}
```
